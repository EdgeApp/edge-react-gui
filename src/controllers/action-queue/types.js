// @flow
/* eslint-disable no-use-before-define */

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'exchange-buy'
  | 'exchange-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type ActionOpExecStatus = 'pending' | 'active' | 'done' | Error

export type ActionOp =
  | {
      type: 'seq',
      actions: ActionOp[]
    }
  | {
      type: 'par',
      actions: ActionOp[]
    }
  | {
      type: 'broadcast-tx',
      pluginId: string,
      rawTx: Uint8Array
    }
  | {
      type: 'exchange-buy',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'exchange-sell',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'loan-borrow',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-deposit',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-repay',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-withdraw',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'swap',
      fromWalletId: string,
      toWalletId: string,
      fromTokenId?: string,
      toTokenId?: string,
      nativeAmount: string,
      amountFor: 'from' | 'to'
    }
  // Useful for development/testing
  | {
      type: 'toast',
      message: string
    }
  | {
      type: 'delay',
      ms: number
    }

//
// Action (After) Effects
//

export type ActionEffect =
  | {
      type: 'seq',
      opIndex: number,
      childEffect: ActionEffect
    }
  | {
      type: 'par',
      childEffects: ActionEffect[]
    }
  | {
      type: 'address-balance',
      address: string,
      aboveAmount?: string,
      belowAmount?: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'tx-confs',
      txId: string,
      walletId: string,
      confirmations: number
    }
  | {
      type: 'price-level',
      currencyPair: string,
      aboveRate?: number,
      belowRate?: number
    }
  | {
      type: 'done',
      error?: Error
    }
  // Useful for development/testing
  | {
      type: 'unixtime',
      timestamp: number
    }
  | {
      type: 'init'
    }

//
// Action Program
//

// On Disk:
export type ActionProgram = {
  programId: string,
  actionOp: ActionOp
}

export type ActionProgramState = {
  programId: string,
  effect?: ActionEffect
}

//
// Internal Types
//

export type ActionQueueItem = {
  program: ActionProgram,
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

export type ExecutionResult = {
  effect: ActionEffect,
  action?: ActionOp // For dryrun
}
export type ExecutionResults = {
  nextState: ActionProgramState
}

//
// Aciton Display API
//

export type ActionDisplayInfo = {
  title: string,
  message: string,
  status: ActionOpExecStatus,
  steps: ActionDisplayInfo[]
}
