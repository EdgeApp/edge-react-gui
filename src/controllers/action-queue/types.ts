/* eslint-disable no-use-before-define */

import { EdgeAccount, EdgeNetworkFee, EdgeTransaction } from 'edge-core-js'

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'wyre-buy'
  | 'wyre-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type SeqActionOp = {
  type: 'seq'
  actions: ActionOp[]
}
export type ParActionOp = {
  type: 'par'
  actions: ActionOp[]
}
export type BroadcastTxActionOp = {
  type: 'broadcast-tx'
  pluginId: string
  rawTx: Uint8Array
}
export type WyreBuyActionOp = {
  type: 'wyre-buy'
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type WyreSellActionOp = {
  type: 'wyre-sell'
  wyreAccountId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanBorrowActionOp = {
  type: 'loan-borrow'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanDepositActionOp = {
  type: 'loan-deposit'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanRepayActionOp = {
  type: 'loan-repay'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanWithdrawActionOp = {
  type: 'loan-withdraw'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type SwapActionOp = {
  type: 'swap'
  amountFor: 'from' | 'to'
  fromTokenId?: string
  fromWalletId: string
  nativeAmount: string
  toTokenId?: string
  toWalletId: string
}
export type ActionOp =
  | SeqActionOp
  | ParActionOp
  | BroadcastTxActionOp
  | WyreBuyActionOp
  | WyreSellActionOp
  | LoanBorrowActionOp
  | LoanDepositActionOp
  | LoanRepayActionOp
  | LoanWithdrawActionOp
  | SwapActionOp

//
// Action (After) Effects
//

export type SeqEffect = {
  type: 'seq'
  opIndex: number
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type ParEffect = {
  type: 'par'
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type AddressBalanceEffect = {
  type: 'address-balance'
  address: string
  aboveAmount?: string
  belowAmount?: string
  walletId: string
  tokenId?: string
}
export type PushEventEffect = {
  type: 'push-event'
  eventId: string
}
export type PriceLevelEffect = {
  type: 'price-level'
  currencyPair: string
  aboveRate?: number
  belowRate?: number
}
export type TxConfsEffect = {
  type: 'tx-confs'
  txId: string
  walletId: string
  confirmations: number
}
export type DoneEffect = {
  type: 'done'
  error?: Error
  cancelled?: boolean
}

export type ActionEffect = SeqEffect | ParEffect | AddressBalanceEffect | PushEventEffect | PriceLevelEffect | TxConfsEffect | DoneEffect

//
// Action Program
//

// Storage:
export type ActionProgram = {
  programId: string
  actionOp: ActionOp
  // Development mode flag
  mockMode?: boolean
}
export type ActionProgramState = {
  clientId: string
  programId: string
  effect?: ActionEffect

  // Flags:
  effective: boolean // Whether the effect is observed
  executing: boolean // Whether the program is executing

  lastExecutionTime: number // The time when the effect was checked
  nextExecutionTime: number // The next time when the effect should be checked again
}

export type ActionQueueItem = {
  program: ActionProgram
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

// Runtime:
export type BroadcastTx = {
  walletId: string
  networkFee: EdgeNetworkFee
  tx: EdgeTransaction
}
export type ExecutableAction = {
  dryrunOutput: ExecutionOutput | null
  execute: () => Promise<ExecutionOutput>
}
export type ExecutionContext = {
  account: EdgeAccount
  clientId: string
}
export type ExecutionOutput = {
  effect: ActionEffect
  broadcastTxs: BroadcastTx[]
}
export type ExecutionResults = {
  nextState: ActionProgramState
}
export type PendingTxMap = {
  [walletId: string]: EdgeTransaction[]
}

//
// Aciton Display API
//

export type ActionDisplayStatus = 'pending' | 'active' | 'done' | Error

export type ActionDisplayInfo = {
  title: string
  message: string
  status: ActionDisplayStatus
  steps: ActionDisplayInfo[]
}
