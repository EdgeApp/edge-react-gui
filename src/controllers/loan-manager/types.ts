import { BorrowEngine, BorrowPlugin } from '../../plugins/borrow-plugins/types'
import { LoanProgramEdge } from './store'

export type LoanAccount = {
  id: string
  borrowPlugin: BorrowPlugin
  borrowEngine: BorrowEngine

  // Action Program Relationships
  programEdges: LoanProgramEdge[]
}

export type LoanAccountMap = {
  [walletId: string]: LoanAccount
}

export type LoanManager = {
  loanAccounts: LoanAccount[]
}
