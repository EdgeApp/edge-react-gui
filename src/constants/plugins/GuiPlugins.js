// @flow

import { Platform } from 'react-native'
import RNFS from 'react-native-fs'

import { type GuiPlugin, type GuiPluginRow } from '../../types/GuiPluginTypes.js'

const hostedUri = Platform.OS === 'android' ? 'file:///android_asset/plugins/' : `file:///${RNFS.MainBundlePath}/plugins/`

export const guiPlugins: { [pluginId: string]: GuiPlugin } = {
  libertyx: {
    pluginId: 'libertyx',
    storeId: 'com.libertyx',
    baseUri: 'https://libertyx.com/a',
    displayName: 'LibertyX',
    originWhitelist: ['https://libertyx.com'],
    permissions: ['location']
  },
  moonpay: {
    pluginId: 'moonpay',
    storeId: 'io.moonpay.buy',
    baseUri: 'https://buy.moonpay.io',
    baseQuery: { apiKey: 'pk_live_Y1vQHUgfppB4oMEZksB8DYNQAdA4sauy', redirectURL: 'https://developer.edge.app/content/conversiontracker.html' },
    queryPromoCode: 'apiKey',
    displayName: 'MoonPay',
    permissions: ['camera'],
    mandatoryPermissions: true
  },
  safello: {
    pluginId: 'safello',
    storeId: 'com.safello',
    baseUri: 'https://safello.com/edge',
    displayName: 'Safello',
    originWhitelist: ['https://safello.com', 'https://app.safello.com', 'http://safello.com']
  },
  'safello-sell': {
    pluginId: 'safello-sell',
    storeId: 'com.safello',
    baseUri: 'https://app.safello.com',
    displayName: 'Safello',
    originWhitelist: ['https://safello.com', 'https://app.safello.com', 'http://safello.com']
  },
  bitsofgold: {
    pluginId: 'bitsofgold',
    storeId: 'bitsofgold',
    baseUri: 'https://www.bitsofgold.co.il',
    queryPromoCode: 'promo_code',
    displayName: 'Bits of Gold',
    permissions: ['camera']
  },
  banxa: {
    pluginId: 'banxa',
    storeId: 'banxa',
    baseUri: 'https://edge.banxa.com',
    displayName: 'Banxa',
    permissions: ['camera']
  },
  simplex: {
    pluginId: 'simplex',
    storeId: 'co.edgesecure.simplex',
    baseUri: hostedUri + 'co.edgesecure.simplex/index.html',
    lockUriPath: true,
    displayName: 'Simplex'
  },
  wyre: {
    pluginId: 'wyre',
    storeId: 'co.edgesecure.wyre',
    baseUri: hostedUri + 'co.edgesecure.wyre/index.html',
    lockUriPath: true,
    displayName: 'Wyre',
    permissions: ['camera']
    // supportEmail: 'support@sendwyre.com'
  },
  bity: {
    pluginId: 'bity',
    storeId: 'com.bity',
    baseUri: hostedUri + 'com.bity/index.html',
    lockUriPath: true,
    needsCountryCode: true,
    queryPromoCode: 'client_value',
    displayName: 'Bity'
    // supportÃŸEmail: 'support@bity.com'
  },
  bitrefill: {
    pluginId: 'bitrefill',
    storeId: 'co.edgesecure.bitrefill',
    baseUri: 'https://embed.bitrefill.com/?ref=nUqaI7Qe&theme=dark&paymentMethods=bitcoin,ethereum,dogecoin,litecoin,dash',
    lockUriPath: true,
    displayName: 'Bitrefill'
  },
  transak: {
    pluginId: 'transak',
    storeId: 'transak',
    baseUri: 'https://global.transak.com',
    baseQuery: { apiKey: '07c66ef7-33e4-47dd-b036-c8d28a50d962', themeColor: '0D2145', disableWalletAddressForm: 'true' },
    displayName: 'Transak',
    permissions: ['camera']
  },
  bitaccess: {
    pluginId: 'bitaccess',
    storeId: 'bitaccess',
    baseUri: 'https://edge.bitaccessbtm.com',
    displayName: 'Bitaccess',
    permissions: ['location', 'camera']
  },
  custom: {
    pluginId: 'custom',
    storeId: 'custom',
    baseUri: '',
    displayName: 'Custom Plugin',
    permissions: ['camera', 'location']
  }
}

export const customPluginRow: GuiPluginRow = {
  pluginId: 'custom',
  deepPath: '',
  deepQuery: {},

  title: 'Custom Dev',
  description: '',
  partnerIconPath: undefined,
  paymentTypeLogoKey: 'credit',
  paymentTypes: [],
  cryptoCodes: []
}
