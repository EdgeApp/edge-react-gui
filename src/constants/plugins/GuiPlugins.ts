import { Platform } from 'react-native'
import RNFS from 'react-native-fs'

import { creditCardPlugin } from '../../plugins/gui/creditCardPlugin'
import { GuiPlugin, GuiPluginRow } from '../../types/GuiPluginTypes'

const hostedUri = Platform.OS === 'android' ? 'file:///android_asset/plugins/' : `file:///${RNFS.MainBundlePath}/plugins/`

export const guiPlugins: { [pluginId: string]: GuiPlugin } = {
  libertyx: {
    pluginId: 'libertyx',
    storeId: 'com.libertyx',
    baseUri: 'https://libertyx.com/a',
    displayName: 'LibertyX',
    originWhitelist: ['https://libertyx.com'],
    permissions: ['location']
  },
  moonpay: {
    pluginId: 'moonpay',
    storeId: 'io.moonpay.buy',
    baseUri: 'https://sell.moonpay.io',
    baseQuery: { apiKey: 'pk_live_Y1vQHUgfppB4oMEZksB8DYNQAdA4sauy', paymentMethod: 'ach_bank_transfer' },
    queryPromoCode: 'apiKey',
    displayName: 'MoonPay',
    permissions: ['camera'],
    mandatoryPermissions: true,
    fixCurrencyCodes: {
      BAT: { pluginId: 'ethereum', tokenId: '0d8775f648430679a709e98d2b0cb6250d2887ef' },
      BCH: { pluginId: 'bitcoincash' },
      BNB: { pluginId: 'binance' },
      BTC: { pluginId: 'bitcoin' },
      CELO: { pluginId: 'celo' },
      CHZ: { pluginId: 'ethereum', tokenId: '3506424f91fd33084466f402d5d97f05f8e3b4af' },
      COMP: { pluginId: 'ethereum', tokenId: 'c00e94cb662c3520282e6f5717214004a7f26888' },
      DAI: { pluginId: 'ethereum', tokenId: '6b175474e89094c44da98b954eedeac495271d0f' },
      DASH: { pluginId: 'dash' },
      DGB: { pluginId: 'digibyte' },
      DOGE: { pluginId: 'dogecoin' },
      DOT: { pluginId: 'polkadot' },
      EOS: { pluginId: 'eos' },
      ETC: { pluginId: 'ethereumclassic' },
      ETH: { pluginId: 'ethereum' },
      FLOW: { pluginId: 'flow' },
      HBAR: { pluginId: 'hedera' },
      LTC: { pluginId: 'litecoin' },
      MATIC_POLYGON: { pluginId: 'polygon' },
      QTUM: { pluginId: 'qtum' },
      RVN: { pluginId: 'ravencoin' },
      TUSD: { pluginId: 'ethereum', tokenId: '0000000000085d4780b73119b644ae5ecd22b376' },
      USDC: { pluginId: 'ethereum', tokenId: 'a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48' },
      USDT: { pluginId: 'ethereum', tokenId: 'dac17f958d2ee523a2206206994597c13d831ec7' },
      XLM: { pluginId: 'stellar' },
      XRP: { pluginId: 'ripple' },
      XTZ: { pluginId: 'tezos' },
      ZRX: { pluginId: 'ethereum', tokenId: 'e41d2489571d322189246dafa5ebde1f4699f498' }
    }
  },
  bitsofgold: {
    pluginId: 'bitsofgold',
    storeId: 'bitsofgold',
    baseUri: 'https://www.bitsofgold.co.il',
    queryPromoCode: 'promo_code',
    displayName: 'Bits of Gold',
    permissions: ['camera']
  },
  banxa: {
    pluginId: 'banxa',
    storeId: 'banxa',
    baseUri: 'https://edge.banxa.com',
    displayName: 'Banxa',
    permissions: ['camera']
  },
  creditcard: {
    pluginId: 'creditcard',
    storeId: '',
    baseUri: '',
    lockUriPath: true,
    nativePlugin: creditCardPlugin,
    displayName: 'Credit Card'
  },
  iach: {
    pluginId: 'creditcard',
    storeId: '',
    baseUri: '',
    lockUriPath: true,
    nativePlugin: creditCardPlugin,
    displayName: 'ACH Bank Transfer'
  },
  simplex: {
    pluginId: 'simplex',
    storeId: 'co.edgesecure.simplex',
    baseUri: hostedUri + 'co.edgesecure.simplex/index.html',
    lockUriPath: true,
    displayName: 'Simplex'
  },
  bity: {
    pluginId: 'bity',
    storeId: 'com.bity',
    baseUri: hostedUri + 'com.bity/index.html',
    lockUriPath: true,
    needsCountryCode: true,
    queryPromoCode: 'client_value',
    displayName: 'Bity'
    // supportÃŸEmail: 'support@bity.com'
  },
  bitrefill: {
    pluginId: 'bitrefill',
    storeId: 'co.edgesecure.bitrefill',
    baseUri: 'https://embed.bitrefill.com/?ref=nUqaI7Qe&theme=dark&paymentMethods=bitcoin,ethereum,dogecoin,litecoin,dash',
    lockUriPath: true,
    displayName: 'Bitrefill'
  },
  bitaccess: {
    pluginId: 'bitaccess',
    storeId: 'bitaccess',
    baseUri: 'https://edge.bitaccessbtm.com',
    displayName: 'Bitaccess',
    permissions: ['location', 'camera']
  },
  ionia: {
    pluginId: 'ionia',
    storeId: 'ioniarewards',
    baseUri: 'https://ionia-edge.web.app',
    displayName: 'Ionia',
    permissions: [] // TODO: need permissions?
  },
  xanpool: {
    pluginId: 'xanpool',
    storeId: 'xanpool',
    baseUri: 'https://widget.xanpool.com',
    baseQuery: { apiKey: 'ae524a0144ccd8dc087af39eabb7a02a', isWebView: 'true' },
    displayName: 'Xanpool',
    permissions: ['camera']
  },
  // Partner whitelabel plugins
  coinhub: {
    pluginId: 'coinhub',
    storeId: 'coinhub',
    baseUri: 'https://coinhubatm.app',
    displayName: 'Coinhub ATMs',
    permissions: ['location']
  },
  custom: {
    pluginId: 'custom',
    storeId: 'custom',
    baseUri: '',
    displayName: 'Custom Plugin',
    permissions: ['camera', 'location']
  }
}

export const IONIA_SUPPORTED_FIATS = ['USD']

export const customPluginRow: GuiPluginRow = {
  pluginId: 'custom',
  deepPath: '',
  deepQuery: {},

  title: 'Custom Dev',
  description: '',
  partnerIconPath: undefined,
  paymentTypeLogoKey: 'paynow',
  paymentTypes: [],
  cryptoCodes: []
}
