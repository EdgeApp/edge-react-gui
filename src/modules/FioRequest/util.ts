export const isRejectedFioRequest = (status: string): boolean => {
  return status === 'rejected'
}
export const isSentFioRequest = (status: string): boolean => {
  return status === 'sent_to_blockchain'
}
