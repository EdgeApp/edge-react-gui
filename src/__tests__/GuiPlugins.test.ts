import { describe, expect, it } from '@jest/globals'

import buyPluginJsonRaw from '../constants/plugins/buyPluginList.json'
import { guiPlugins } from '../constants/plugins/GuiPlugins'
import sellPluginJsonRaw from '../constants/plugins/sellPluginList.json'
import { asGuiPluginJson, GuiPlugin, GuiPluginRow } from '../types/GuiPluginTypes'
import { filterGuiPluginJson, makePluginUri } from '../util/GuiPluginTools'

const buyPluginJson = asGuiPluginJson(buyPluginJsonRaw)
const sellPluginJson = asGuiPluginJson(sellPluginJsonRaw)

describe('Production plugin data', () => {
  it('Buy & sell plugins have real pluginId fields', () => {
    for (const row of [...buyPluginJson, ...sellPluginJson]) {
      if (typeof row === 'string') continue
      const { pluginId } = row
      if (pluginId != null) expect(guiPlugins[pluginId]).not.toBeUndefined()
    }
  })

  it('Buy plugins match snapshot on iOS + US', () => {
    expect(filterGuiPluginJson(buyPluginJson, 'ios', 'US', {})).toMatchSnapshot()
  })

  it('Sell plugins match snapshot on iOS + US', () => {
    expect(filterGuiPluginJson(sellPluginJson, 'ios', 'US', {})).toMatchSnapshot()
  })

  it('Buy plugins match snapshot on android + IL', () => {
    expect(filterGuiPluginJson(buyPluginJson, 'android', 'IL', {})).toMatchSnapshot()
  })
})

describe('GuiPlugins tools', () => {
  it('filter android + US as expected', () => {
    const list = filterGuiPluginJson(testJson, 'android', 'US', {})

    expect(titles(list)).toEqual(['Credit card', 'Wire transfer'])
    expect(pluginIds(list)).toEqual(['phony', 'gox'])
    expect(list).toMatchSnapshot()
  })

  it('filter ios + US as expected', () => {
    const list = filterGuiPluginJson(testJson, 'ios', 'US', {})

    expect(titles(list)).toEqual(['Apple Pay', 'Wire transfer', 'Credit card'])
    expect(pluginIds(list)).toEqual(['phony', 'gox', 'phony'])
    expect(list).toMatchSnapshot()
  })

  it('filter ios + US - gox as expected', () => {
    const list = filterGuiPluginJson(testJson, 'ios', 'US', { gox: true })

    expect(titles(list)).toEqual(['Apple Pay', 'Credit card'])
    expect(pluginIds(list)).toEqual(['phony', 'phony'])
    expect(list).toMatchSnapshot()
  })

  it('filter ios + JP as expected', () => {
    const list = filterGuiPluginJson(testJson, 'ios', 'JP', {})

    expect(titles(list)).toEqual(['Nice lawsuit'])
    expect(pluginIds(list)).toEqual(['gox'])
    expect(list).toMatchSnapshot()
  })

  it('filter everything for GB', () => {
    const list = filterGuiPluginJson(testJson, 'ios', 'GB', {})

    expect(titles(list)).toEqual([])
  })

  it("produce correct URI's", () => {
    const testPlugin: GuiPlugin = {
      pluginId: 'local',
      storeId: 'custom',
      baseUri: 'file://test/',
      baseQuery: { api_key: 'edge' },
      displayName: 'Test Plugin'
    }
    const opts = {
      deepPath: 'sell',
      deepQuery: { kickback: null },
      promoCode: 'deals'
    }

    expect(makePluginUri(testPlugin, opts)).toEqual('file://test/sell?api_key=edge&kickback')
    expect(makePluginUri({ ...testPlugin, lockUriPath: true }, opts)).toEqual('file://test/?api_key=edge&kickback')
    expect(makePluginUri({ ...testPlugin, queryPromoCode: 'cheat' }, opts)).toEqual('file://test/sell?api_key=edge&kickback&cheat=deals')
  })
})

function pluginIds(plugins: GuiPluginRow[]): string[] {
  return plugins.map(plugin => plugin.pluginId)
}

function titles(plugins: GuiPluginRow[]): string[] {
  return plugins.map(plugin => plugin.title)
}

// This test data is designed to exercise specific problem areas:
const testJson = asGuiPluginJson([
  {
    id: 'phony-credit',
    pluginId: 'phony',
    title: 'Credit card',
    paymentTypes: ['credit'],
    forCountries: ['US', 'AU']
  },
  {
    id: 'phony-apple',
    pluginId: 'phony', // Duplicated plugin
    deepPath: '/applePay', // Different URL
    title: 'Apple Pay',
    paymentTypes: ['applepay'],
    forCountries: ['US'],
    forPlatform: 'ios'
  },
  {
    id: 'gox',
    pluginId: 'gox',
    title: 'Wire transfer',
    paymentTypes: ['wire', 'hack'],
    forCountries: ['US', 'JP']
  },
  '----- Sorting -----',
  {
    id: 'phony-credit',
    sortIndex: 3
  },
  {
    id: 'gox',
    sortIndex: 4
  },
  '----- Platform Specifics -----',
  {
    id: 'phony-apple',
    forPlatform: 'ios',
    sortIndex: 2 // Apple Pay to top on iOS
  },
  {
    id: 'phony-credit',
    forPlatform: 'ios',
    sortIndex: 5 // Credit to bottom on iOS
  },
  '----- Country Specifics -----',
  {
    id: 'gox',
    title: 'Nice lawsuit', // Special title in Japan
    forCountries: ['JP', 'GB'], // Note that GB isn't supported above
    sortIndex: 1
  }
])
