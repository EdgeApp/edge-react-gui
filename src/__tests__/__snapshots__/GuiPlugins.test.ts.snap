// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuiPlugins tools filter android + US as expected 1`] = `
[
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "phony",
    "title": "Credit card",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "wire",
      "hack",
    ],
    "pluginId": "gox",
    "title": "Wire transfer",
  },
]
`;

exports[`GuiPlugins tools filter ios + JP as expected 1`] = `
[
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "wire",
      "hack",
    ],
    "pluginId": "gox",
    "title": "Nice lawsuit",
  },
]
`;

exports[`GuiPlugins tools filter ios + US - gox as expected 1`] = `
[
  {
    "cryptoCodes": [],
    "deepPath": "/applePay",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "applepay",
    ],
    "pluginId": "phony",
    "title": "Apple Pay",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "phony",
    "title": "Credit card",
  },
]
`;

exports[`GuiPlugins tools filter ios + US as expected 1`] = `
[
  {
    "cryptoCodes": [],
    "deepPath": "/applePay",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "applepay",
    ],
    "pluginId": "phony",
    "title": "Apple Pay",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "wire",
      "hack",
    ],
    "pluginId": "gox",
    "title": "Wire transfer",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "phony",
    "title": "Credit card",
  },
]
`;

exports[`Production plugin data Buy plugins match snapshot on android + IL 1`] = `
[
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "Fee: 5-7%
Settlement: 10 - 30 minutes",
    "paymentType": "googlepay",
    "paymentTypeLogoKey": "googlepay",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "creditcard",
    "title": "Google Pay",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "Fee: 5-7%
Settlement: 10 - 30 minutes",
    "paymentType": "credit",
    "paymentTypeLogoKey": "credit",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "creditcard",
    "title": "Credit and Debit Card",
  },
  {
    "cryptoCodes": [
      "BTC",
      "ETH",
    ],
    "deepPath": "/order/buy",
    "deepQuery": {
      "order_id": "null",
      "page": "0",
    },
    "description": "Fee: 5.95%
Settlement: 1 - 48 hours",
    "partnerIconPath": "bits-of-gold-logo-sm.png",
    "paymentTypeLogoKey": "bank",
    "paymentTypes": [
      "bank",
    ],
    "pluginId": "bitsofgold",
    "title": "Israeli Bank Transfer",
  },
]
`;

exports[`Production plugin data Buy plugins match snapshot on iOS + US 1`] = `
[
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "Fee: 4-5%
Settlement: 1 - 4 days",
    "paymentType": "iach",
    "paymentTypeLogoKey": "bank",
    "paymentTypes": [
      "iach",
    ],
    "pluginId": "iach",
    "title": "ACH Bank Transfer",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "Fee: 5-7%
Settlement: 10 - 30 minutes",
    "paymentType": "applepay",
    "paymentTypeLogoKey": "applepay",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "creditcard",
    "title": "Apple Pay",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "Fee: 5-7%
Settlement: 10 - 30 minutes",
    "paymentType": "credit",
    "paymentTypeLogoKey": "credit",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "creditcard",
    "title": "Credit and Debit Card",
  },
  {
    "cryptoCodes": [
      "BTC",
    ],
    "deepPath": "/",
    "deepQuery": {},
    "description": "Fee: 3-8%
Settlement: Instant",
    "partnerIconPath": "libertyXlogo.png",
    "paymentTypeLogoKey": "cash",
    "paymentTypes": [
      "cash",
      "credit",
    ],
    "pluginId": "libertyx",
    "title": "In-person Cash and Debit Card",
  },
]
`;

exports[`Production plugin data Sell plugins match snapshot on iOS + US 1`] = `
[
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "Fee: 4-5% / Settlement: 1 - 4 days",
    "partnerIconPath": "moonpay_symbol_prp.png",
    "paymentTypeLogoKey": "bank",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "moonpay",
    "title": "ACH Bank Transfer",
  },
  {
    "cryptoCodes": [],
    "deepPath": "",
    "deepQuery": {},
    "description": "Buy prepaid, single-load Visa® cards
Fee: 0% / Settlement: Instant
Limit: $1,000 per card, $10,000 per day
",
    "paymentType": "credit",
    "paymentTypeLogoKey": "visa",
    "paymentTypes": [
      "credit",
    ],
    "pluginId": "rewardscard",
    "title": "Visa® Card Program",
  },
  {
    "cryptoCodes": [
      "BTC",
      "DASH",
      "DOGE",
      "ETH",
      "LTC",
    ],
    "deepPath": "",
    "deepQuery": {},
    "description": "Buy gift cards and reload phones with crypto",
    "partnerIconPath": "bitrefill-2.png",
    "paymentTypeLogoKey": "giftcard",
    "paymentTypes": [
      "giftcard",
    ],
    "pluginId": "bitrefill",
    "title": "Gift Cards",
  },
]
`;
